src/meili/match_route.cc

```diff
@@ -208,69 +216,152 @@ bool MergeRoute(std::vector<EdgeSegment>& route, const State& source, const Stat
+      return false;
+    }
+    // Make a trivial route
+    segments.emplace_back(source_edge.id, source_edge.percent_along, source_edge.percent_along);*/
+  }
+
+  route.insert(route.end(), segments.crbegin(), segments.crend());
   return true;
 }
 
-std::vector<EdgeSegment> MergeRoute(const State& source, const State& target) {
-  std::vector<EdgeSegment> route;
-  MergeRoute(route, source, target);
-  return route;
+void cut_segments(const std::vector<MatchResult>& match_results,
+                  int first_idx,
+                  int last_idx,
+                  const std::vector<EdgeSegment>& segments,
+                  std::vector<EdgeSegment>& new_segments) {
+  auto first_segment = segments.begin();
+  int prev_idx = first_idx;
+  for (int curr_idx = first_idx + 1; curr_idx <= last_idx; ++curr_idx) {
+    const MatchResult& curr_match = match_results[curr_idx];
+    const MatchResult& prev_match = match_results[prev_idx];
+
+    // skip if we do not need to cut
+    if (!curr_match.is_break_point && curr_idx != last_idx) {
+      continue;
+    }
+    // we want to handle to loop by locating the correct target edge by comparing the distance alone
+    bool loop = prev_match.edgeid == curr_match.edgeid &&
+                prev_match.distance_along > curr_match.distance_along;
+    // if it is a loop, we start the search after the first edge
+    auto last_segment = std::find_if(first_segment + static_cast<size_t>(loop), segments.end(),
+                                     [&curr_match](const EdgeSegment& segment) {
+                                       return (segment.edgeid == curr_match.edgeid);
+                                     });
+    assert(last_segment != segments.cend());
+
+    // we need to close the previous edge
+    size_t old_size = new_segments.size();
+    new_segments.insert(new_segments.cend(), first_segment, last_segment + 1);
+    new_segments[old_size].first_match_idx = prev_idx;
+    // when the points got interpolated, we want to use the match results' distance along
+    // otherwise, we use the segment's source or target because if it is a node snap, the
+    // match result can only hold one candidate, we need either side of the node.
+    new_segments[old_size].source =
+        prev_match.HasState() ? first_segment->source : prev_match.distance_along;
+    new_segments.back().last_match_idx = curr_idx;
+    new_segments.back().target =
+        curr_match.HasState() ? last_segment->target : curr_match.distance_along;
+
+    first_segment = last_segment;
+    prev_idx = curr_idx;
+  }
 }
 
-template <typename match_iterator_t>
-std::vector<EdgeSegment>
-ConstructRoute(const MapMatcher& mapmatcher, match_iterator_t begin, match_iterator_t end) {
-  if (begin == end) {
+std::vector<EdgeSegment> ConstructRoute(const MapMatcher& mapmatcher,
+                                        const std::vector<MatchResult>& match_results,
+                                        baldr::GraphReader& reader) {
```

Comment:
What are our conventions around using smart pointers vs bare pointers here? If I remember correctly, in some places in meili we're using `shared_ptr` but in other places we're not (like right here). We should probably be using smart points consistently throughout.

/cc @yuzheyan @kevinkreiser

Reply:
was this meant for another place in the diff? i cant tell what you are refering to here as i dont see any pointers in the function signature.

Reply:
I'm referring to 'baldr::GraphReader&', which is a reference, but not a pointer (sorry misspoke here). It still refers to an underlying resource that we want consistent memory semantics around, right?

Reply:
can you be more specific? are you suggesting something in place of a graphreader reference? it needs to be a reference because the reader is essentially a cache. we could technically create a new one but its not performant to do so. in this case, we aquire the resource outside of this class and pass it down into the class. basically meili doesnt ever own its graphreader, it only borrows it by reference from thor who owns it.

Reply:
we clearified this offline. basically the reason a few of the interfaces pass around shared_ptr of graph reader was to get around being able to either make one or use an already made one. since this function doesnt need to ever own it it doesnt require any more scoping